File: leetcode.com/python/199_Binary_Tree_Right_Side_View.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if not root:
            return []
        rightView = []
        queue = [root, ]
        currentLevelLength = 0
        while queue:
            currentLevelLength = len(queue)
            for idx in range(currentLevelLength):
                node = queue.pop(0)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
                if idx == (currentLevelLength - 1):
                    rightView.append(node.val)
        return rightView



#  My solution during mock contest
from collections import deque
class Solution(object):
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        queue = deque([root])
        rightView = []
        while queue:
            currentLevelLen = len(queue)
            for i in range(currentLevelLen):
                currentNode = queue.popleft()
                if currentNode:
                    if i == currentLevelLen - 1:
                        rightView.append(currentNode.val)
                    if currentNode.left:
                        queue.append(currentNode.left)
                    if currentNode.right:
                        queue.append(currentNode.right)
        return rightView

File: leetcode.com/swift/DS_ALGO_PRAC.playground/Sources/swift/199_Binary_Tree_Right_Side_View.swift
import Foundation
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */

// DFS
class Solution {
    func rightSideView(_ root: TreeNode?) -> [Int] {
        guard var root = root else {
            return []
        }
        var rightView = [Int]()
        var (maxDepthSoFar, currentDepth) = (-1, -1)
        rightSideViewDFSHelper(root, currentDepth + 1, &maxDepthSoFar, &rightView)
        return rightView
    }
    
    private func rightSideViewDFSHelper(_ root: TreeNode?, _ currentDepth: Int, _ maxDepthSoFar: inout Int, _ rightView: inout [Int]) {
        guard var root = root else {
            return
        }
        if currentDepth > maxDepthSoFar {
            rightView.append(root.val)
            maxDepthSoFar = currentDepth
        }
        rightSideViewDFSHelper(root.right, currentDepth + 1, &maxDepthSoFar, &rightView)
        rightSideViewDFSHelper(root.left, currentDepth + 1, &maxDepthSoFar, &rightView)
    }
}


// BFS with level size measurement
class Solution {
    func rightSideView(_ root: TreeNode?) -> [Int] {
        guard let root = root else {
            return []
        }
        var result = [Int](), queue = [TreeNode](), currentLevelLength = 1
        queue.append(root)
        while !queue.isEmpty {
            var rightMostNode = Int.min, currentLevelLength = queue.count
            while currentLevelLength > 0 {
                let node = queue.removeFirst()
                rightMostNode = node.val
                currentLevelLength -= 1
                if let left = node.left {
                    queue.append(left)
                }
                if let right = node.right {
                    queue.append(right)
                }
            }
            result.append(rightMostNode)
        }
        return result
    }
}
